ga4gh/protocol.py

```diff
@@ -102,57 +79,69 @@ def toJSONDict(self):
         """
         Returns a JSON dictionary representation of this ProtocolElement.
         """
-        # TODO this is horrible! Need to streamline this and reuse some
-        # of the code in _decode below
-        return json.loads(json.dumps(self, cls=ProtocolElementEncoder))
+        out = {}
+        for field in self.schema.fields:
+            val = self.__dict__[field.name]
+            if self.isEmbeddedType(field.name):
+                if isinstance(val, list):
+                    out[field.name] = list(el.toJSONDict() for el in val)
+                elif val is None:
+                    out[field.name] = None
+                else:
+                    out[field.name] = val.toJSONDict()
+            elif isinstance(val, list):
+                out[field.name] = list(val)
+            else:
+                out[field.name] = val
+        return out
 
     @classmethod
     def validate(cls, jsonDict):
         """
-        Validates the specified json dictionary to determine if it is an
+        Validates the specified JSON dictionary to determine if it is an
         instance of this element's schema.
         """
         return avro.io.validate(cls.schema, jsonDict)
 
     @classmethod
-    def fromJSON(cls, jsonStr):
+    def fromJSONString(cls, jsonStr):
         """
         Returns a decoded ProtocolElement from the specified JSON string.
-
-        TODO change the signature of this to take a json dictionary, and the
-        fold in _decode in here.
         """
         d = json.loads(jsonStr)
-        return cls._decode(d)
+        return cls.fromJSONDict(d)
 
     @classmethod
-    def _decode(cls, d):
-        # TODO This code is needlessly obscure and needs refactoring.
-        instance = cls()
-        schema = cls.schema
+    def fromJSONDict(cls, d):
```

Comment:
I'm good with that. I wasn't sure how we came down on the "v for variant" sorts of things.